Implement consent creation, retrieval and revocation

AthulyaWeerakoon opened the pull request:
## Implement consent creation, retrieval and revocation.

> This PR imeplements consent creation, retrieval and revocation for the new nextgenpsd2 OpenBanking toolkit

**Issue link:** *required*

**Doc Issue:** *Optional, link issue from [documentation repository](https://github.com/wso2/docs-ob/issues)*

**Applicable Labels:** *Spec, product, version, type (specify requested labels)*

------

### Development Checklist

1. [x] Built complete solution with pull request in place.
2. [ ] Ran checkstyle plugin with pull request in place.
3. [x] Ran Findbugs plugin with pull request in place.
4. [x] Ran FindSecurityBugs plugin and verified report.
5. [ ] Formatted code according to WSO2 code style.
6. [x] Have you verify the PR does't commit any keys, passwords, tokens, usernames, or other secrets?
7. [ ] Migration scripts written (if applicable).
8. [ ] Have you followed secure coding standards in [WSO2 Secure Engineering Guidelines](http://wso2.com/technical-reports/wso2-secure-engineering-guidelines)?

### Testing Checklist

1. [ ] Written unit tests.
2. [ ] Documented test scenarios(link available in guides).
3. [ ] Written automation tests (link available in guides).
4. [ ] Verified tests in multiple database environments (if applicable).
5. [ ] Verified tests in multiple deployed specifications (if applicable).
6. [ ] Tested with OBBI enabled  (if applicable).
7. [ ] Tested with specification regulatory conformance suites  (if applicable).

**Automation Test Details**

| Test Suite        | Test Script IDs   |
| ----------------- | ----------------- |
| Integration Suite | *TCXXXXX, TCXXXX* |

**Conformance Tests Details**

| Test Suite Name  | Test Suite Version | Scenarios  | Result   |
| ---------------- | ------------------ | ---------- | -------- |
| *Security Suite* | *VX.X*             | *Foo, Bar* | *Passed* |

## Resources

**Knowledge Base:** https://sites.google.com/wso2.com/open-banking/

**Guides:** https://sites.google.com/wso2.com/open-banking/developer-guides

AthulyaWeerakoon committed:
Implement consent creation and consent retrieval

Ashi1993 commented on components/reference-implementation-openbanking-nextgenpsd2/src/main/java/org/wso2/openbanking/nextgenpsd2/extensions/api/PreProcessConsentCreationApi.java:
```diff
@@ -34,7 +46,63 @@ public class PreProcessConsentCreationApi {
+            // Extract consent initiation data
+            JSONObject consentInitiationDataJSON;
+            try {
+                consentInitiationDataJSON = CommonConsentValidationUtil.convertObjectToJson(requestBody.getData()
```
Can you explain the reason for this convertion

AthulyaWeerakoon replied:
It's done because the initiation payload is set as an object in the DTO, which needs to be converted to a JSON to process easier.

Ashi1993 commented on components/reference-implementation-openbanking-nextgenpsd2/src/main/java/org/wso2/openbanking/nextgenpsd2/extensions/api/PreProcessConsentCreationApi.java:
```diff
@@ -34,7 +46,63 @@ public class PreProcessConsentCreationApi {
+            if (consentHandler != null) {
+                consentHandler.handleCreation(requestBody, validationResponse);
+            } else {
+                throw new FailedValidationException(FailedValidationException.ErrorCode.NOT_FOUND,
```
Error code and the error message doesn't seems aligned

AthulyaWeerakoon replied:
This was from the old toolkit and I didn't change it because this error is triggered when the resource for the given path is not found.

Ashi1993 commented on components/reference-implementation-openbanking-nextgenpsd2/src/main/java/org/wso2/openbanking/nextgenpsd2/extensions/api/EnrichConsentCreationResponseApi.java:
```diff
@@ -34,7 +46,41 @@ public class EnrichConsentCreationResponseApi {
+                    ErrorUtil.getErrorResponse(ConsentExtensionConstants.INVALID_REQUEST, e.getMessage())
+            ).toString()).build();
+
+        } catch (Exception e) {
```
Why are catching a general exception here?

AthulyaWeerakoon replied:
Fixed, they are no longer caught.

Ashi1993 commented on components/reference-implementation-openbanking-nextgenpsd2/src/main/java/org/wso2/openbanking/nextgenpsd2/extensions/api/PreProcessConsentCreationApi.java:
```diff
@@ -34,7 +46,63 @@ public class PreProcessConsentCreationApi {
+                    ErrorUtil.getErrorResponse(ConsentExtensionConstants.INVALID_REQUEST, e.getMessage())
+            ).toString()).build();
+
+        } catch (Exception e) {
```
Why are catching a general exception here? check all places

AthulyaWeerakoon replied:
Fixed, they are no longer caught.

Ashi1993 commented on components/reference-implementation-openbanking-nextgenpsd2/src/main/java/org/wso2/openbanking/nextgenpsd2/extensions/api/PreProcessConsentCreationApi.java:
```diff
@@ -34,7 +46,63 @@ public class PreProcessConsentCreationApi {
+            log.error("Validation failed for consent creation. Returning failed response.", e);
+            return Response.ok().entity(e.getFormattedErrorAsString()).build();
+
+        } catch (JSONException e) {
```
From which method this exception is thrown?

AthulyaWeerakoon replied:
It used to be thrown from from the handleCreation method, but it no longer does.

Ashi1993 commented on components/reference-implementation-openbanking-nextgenpsd2/src/main/java/org/wso2/openbanking/nextgenpsd2/extensions/exceptions/ServerException.java:
```diff
@@ -0,0 +1,106 @@
+public class ServerException extends RuntimeException {
+
+    public enum ErrorCode {
+        BAD_REQUEST(Response.Status.BAD_REQUEST),
```
Can server exception return a 400

AthulyaWeerakoon replied:
Fixed. Now there's a common `ApiException` class which is extended by `BadRequestException` and `ServerErrorException` which are in turn thrown appropriately.

Ashi1993 commented on components/reference-implementation-openbanking-nextgenpsd2/src/main/java/org/wso2/openbanking/nextgenpsd2/extensions/exceptions/ServerException.java:
```diff
@@ -0,0 +1,106 @@
+    public String getFormattedErrorAsString() {
+        return getFormattedError().toString();
+    }
+}
```
Add EoF line

AthulyaWeerakoon replied:
Added.

Ashi1993 commented on components/reference-implementation-openbanking-nextgenpsd2/src/main/java/org/wso2/openbanking/nextgenpsd2/extensions/api/PreProcessConsentRetrievalApi.java:
```diff
@@ -34,7 +47,47 @@ public class PreProcessConsentRetrievalApi {
-        return Response.ok().entity("magic!").build();
+    public Response preProcessConsentRetrievalPost(@Valid @NotNull PreProcessConsentRequestBody requestBody) {
+        Log log = LogFactory.getLog(PreProcessConsentCreationApi.class);
+        SuccessResponseForResponseAlternation validationResponse = new SuccessResponseForResponseAlternation();
```
Why are we creating the object here and pass to other handlers

AthulyaWeerakoon replied:
Fixed, as discussed in offline review.

AthulyaWeerakoon committed:
Maven checkstyle fixes

AthulyaWeerakoon committed:
Refactor module structure

AthulyaWeerakoon committed:
Refactor consent handler

Refactor consent handler to better reflect its purpose

AthulyaWeerakoon committed:
Refactor handler interfaces

Refactor handler interfaces to better reflect their purpose

AthulyaWeerakoon committed:
Refactor and reorganize module files

Refactored and reorganized module files as suggested in the code review,
- Renamed Shim to Impl
- Combined the two existing handler interfaces
- Split ServerException
- Seperated Handelers, Impl, Constants and Util classes

AthulyaWeerakoon committed:
Add consent revocation

AthulyaWeerakoon committed:
Fix test failures

AthulyaWeerakoon committed:
Refactor ConfigurableProperties

AthulyaWeerakoon committed:
Rename request Id header name

AthulyaWeerakoon committed:
Do validations using the hibernate validator

AthulyaWeerakoon committed:
Improve logging and remove redundant logs

AthulyaWeerakoon committed:
Do payment initiation validations using the Hibernate Validator

AthulyaWeerakoon committed:
Do COF initiation validations using the Hibernate Validator

AthulyaWeerakoon committed:
Fixed server error and bad request error handling

AthulyaWeerakoon committed:
Minor code cleanups

AthulyaWeerakoon committed:
Minor exception handling fix for convertObjectToJson utility method

AthulyaWeerakoon committed:
Refractor handler and api impl class names

VimukthiRajapaksha commented on components/reference-implementation-openbanking-nextgenpsd2/pom.xml:
```diff
@@ -147,6 +147,19 @@
+                    <suppressionsLocation>
+                        https://raw.githubusercontent.com/wso2/code-quality-tools/v1.3/checkstyle/suppressions.xml
+                    </suppressionsLocation>
+                    <excludes>**/generated/**/**,**/RestApplication.java,**/RestResourceRoot.java</excludes>
```
Does this work? we usually follow `<excludes><exclude></exclude></excludes>` pattern

AthulyaWeerakoon replied:
I first went with `<excludes><exclude></exclude></excludes>` but, apparently it's not supported by the checkstyle plugin, but does work with other plugins like JaCoCo. Checkstyle expects regex string (https://maven.apache.org/plugins/maven-checkstyle-plugin/check-mojo.html#excludes)

VimukthiRajapaksha commented on components/reference-implementation-openbanking-nextgenpsd2/pom.xml:
```diff
@@ -147,6 +147,19 @@
     <build>
         <plugins>
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-checkstyle-plugin</artifactId>
```
Can't we move this plugin to the root pom?

AthulyaWeerakoon replied:
It is in the root pom, versioning details and everything, but shouldn't it be included in the child project as well for it to take effect in build? (https://maven.apache.org/pom.html#Plugin_Management)

VimukthiRajapaksha commented on components/reference-implementation-openbanking-nextgenpsd2/src/main/java/org/wso2/openbanking/nextgenpsd2/extensions/api/generated/PreProcessClientCreationApi.java:
```diff
@@ -1,8 +1,8 @@
-package org.wso2.openbanking.nextgenpsd2.extensions.api;
+package org.wso2.openbanking.nextgenpsd2.extensions.api.generated;
```
We need to add license headers. Let's check all classes

AthulyaWeerakoon replied:
To generated classes as well?

VimukthiRajapaksha commented on components/reference-implementation-openbanking-nextgenpsd2/src/main/java/org/wso2/openbanking/nextgenpsd2/extensions/api/generated/PreProcessConsentCreationApi.java:
```diff
@@ -35,6 +36,6 @@ public class PreProcessConsentCreationApi {
     })
     public Response preProcessConsentCreationPost(@Valid @NotNull PreProcessConsentCreationRequestBody preProcessConsentCreationRequestBody) {
-        return Response.ok().entity("magic!").build();
+        return ConsentManagementAPIImpl.preProcessConsentCreation(preProcessConsentCreationRequestBody);
```
Let's follow java naming conventions[1]. Maybe we could rename this to `createPreProcessConsent`,  `preProcessConsent` or something similar

[1] https://www.oracle.com/java/technologies/javase/codeconventions-namingconventions.html

AthulyaWeerakoon replied:
Oh, it was meant as a verb meaning pre-process consent creation, as in pre-process the work of creating a consent.

VimukthiRajapaksha commented on components/reference-implementation-openbanking-nextgenpsd2/src/main/java/org/wso2/openbanking/nextgenpsd2/extensions/api/impl/ConsentManagementAPIImpl.java:
```diff
@@ -0,0 +1,228 @@
+/**
+ * Implementation class maintaining the methods for consent management extension APIs.
+ */
+public class ConsentManagementAPIImpl {
```
Consider renaming this class. `Impl` implies we're implementing an interface.

VimukthiRajapaksha commented on components/reference-implementation-openbanking-nextgenpsd2/src/main/java/org/wso2/openbanking/nextgenpsd2/extensions/api/impl/ConsentManagementAPIImpl.java:
```diff
@@ -0,0 +1,228 @@
+    /**
+     * Method for returning the response for enriching consent creation request.
+     *
+     * @param requestBody
+     * @return
```
Let's explain these values

AthulyaWeerakoon replied:
Added.

VimukthiRajapaksha commented on components/reference-implementation-openbanking-nextgenpsd2/src/main/java/org/wso2/openbanking/nextgenpsd2/extensions/api/impl/ConsentManagementAPIImpl.java:
```diff
@@ -0,0 +1,228 @@
+            SuccessResponseForResponseAlternation validationResponse;
+
+            if (consentHandler != null) {
+                validationResponse = consentHandler.enrichCreationResponse(requestBody);
```
Instead of returning in L72, can't we directly return after this line?

AthulyaWeerakoon replied:
Fixed.

VimukthiRajapaksha commented on components/reference-implementation-openbanking-nextgenpsd2/src/main/java/org/wso2/openbanking/nextgenpsd2/extensions/api/impl/ConsentManagementAPIImpl.java:
```diff
@@ -0,0 +1,228 @@
+                return Response.status(Response.Status.BAD_REQUEST).entity(errorResponse).build();
+            }
+
+            return Response.ok().entity(new JSONObject(validationResponse).toString()).build();
```
Why do we need to manually convert this to a JSONObject here?

AthulyaWeerakoon replied:
Otherwise the toString method for this class returns something like a class description, not exactly a JSON formatted string as expected as it's response.

VimukthiRajapaksha commented on components/reference-implementation-openbanking-nextgenpsd2/src/main/java/org/wso2/openbanking/nextgenpsd2/extensions/api/impl/ConsentManagementAPIImpl.java:
```diff
@@ -0,0 +1,228 @@
+        } catch (BadRequestException | ServerErrorException e) {
+            log.error("[" + requestId + "] " + "An error occurred creating consent.", e);
+            return Response.status(e.getStatus()).entity(e.getFormattedErrorAsString()).build();
+        } catch (org.wso2.openbanking.nextgenpsd2.extensions.exceptions.ApiException e) {
```
Let's move the FQN into the imports section

AthulyaWeerakoon replied:
Oh this line was a mistake, added by intelliJ. It has been fixed since then.

VimukthiRajapaksha commented on components/reference-implementation-openbanking-nextgenpsd2/src/main/java/org/wso2/openbanking/nextgenpsd2/extensions/api/impl/ConsentManagementAPIImpl.java:
```diff
@@ -0,0 +1,228 @@
+            log.error("[" + requestId + "] " + "An error occurred creating consent.", e);
+            return Response.status(e.getStatus()).entity(e.getFormattedErrorAsString()).build();
+        } catch (org.wso2.openbanking.nextgenpsd2.extensions.exceptions.ApiException e) {
+            throw new RuntimeException(e);
```
Don't we need to return a http response here?

AthulyaWeerakoon replied:
Yes we do. This is a code segment accidentally generated by IntelliJ but went unseen in testing because by catching `BadRequestException `and `ServerErrorException `prior to this, this code block is never reached (since `ApiException `is not thrown, only it's children `BadRequestException `and `ServerErrorException `are thrown). It has been fixed since then.

VimukthiRajapaksha commented on components/reference-implementation-openbanking-nextgenpsd2/src/main/java/org/wso2/openbanking/nextgenpsd2/extensions/api/impl/ConsentManagementAPIImpl.java:
```diff
@@ -0,0 +1,228 @@
+            SuccessResponseForResponseAlternation validationResponse;
+
+            if (consentHandler != null) {
+                validationResponse = consentHandler.enrichCreationResponse(requestBody);
```
If we're enriching the requestBody here, don't we need to accept an enrichedRequestBody? If we're validating the requestBody, may be we can rename the method to `validateCreationRequest` ?

AthulyaWeerakoon replied:
This method is for enriching the response, not validating. Validated consent creation response is returned from the `preProcessConsentCreation` method, and is forwarded as the response from the `/pre-process-consent-creation` endpoint. Once the accelerator process that, it forwards the response from that process to the endpoint that calls this method (`/enrich-consent-creation-response`) and this method's responsibility is to make the consent creation response OB specification compliant.

VimukthiRajapaksha commented on components/reference-implementation-openbanking-nextgenpsd2/src/main/java/org/wso2/openbanking/nextgenpsd2/extensions/api/impl/ConsentManagementAPIImpl.java:
```diff
@@ -0,0 +1,228 @@
+                consentInitiationDataJSON = CommonConsentValidationUtil.convertObjectToJson(requestBody.getData()
+                        .getConsentInitiationData());
+
+                if (consentInitiationDataJSON.isEmpty()) {
```
Can't we do a `requestBody.getData() == null` check here?

AthulyaWeerakoon replied:
But that would fail to consider the scenario where the data is `{}` (empty JSON Object).

VimukthiRajapaksha commented on components/reference-implementation-openbanking-nextgenpsd2/src/main/java/org/wso2/openbanking/nextgenpsd2/extensions/api/impl/ConsentManagementAPIImpl.java:
```diff
@@ -0,0 +1,228 @@
+                                    TPPMessage.CodeEnum.FORMAT_ERROR, ErrorConstants.PAYLOAD_NOT_PRESENT_ERROR));
+                    return Response.ok().entity(errorResponse.toString()).build();
+                }
+
```
```suggestion
```

VimukthiRajapaksha commented on components/reference-implementation-openbanking-nextgenpsd2/src/main/java/org/wso2/openbanking/nextgenpsd2/extensions/api/impl/ConsentManagementAPIImpl.java:
```diff
@@ -0,0 +1,228 @@
+                }
+
+            } catch (JSONException e) {
+                // If payload is not JSON
```
Need to log the exception. Let's check other places as well. If needed we can move this logic to a separate method and do the JSON validation there.

AthulyaWeerakoon replied:
Will add.

AthulyaWeerakoon replied:
Added.

VimukthiRajapaksha commented on components/reference-implementation-openbanking-nextgenpsd2/src/main/java/org/wso2/openbanking/nextgenpsd2/extensions/api/impl/ConsentManagementAPIImpl.java:
```diff
@@ -0,0 +1,228 @@
+                }
+
+            } catch (JSONException e) {
+                // If payload is not JSON
```
Doesn't this handles by the `@Consumes({ "application/json" })` annotation in the preProcessConsentCreationAPI class?

AthulyaWeerakoon replied:
To some extent yes, but it would still not throw an error if this initiation object was a string or an array or anything that's not a JSONObject, but is a compatible JSON value.

VimukthiRajapaksha commented on components/reference-implementation-openbanking-nextgenpsd2/src/main/java/org/wso2/openbanking/nextgenpsd2/extensions/api/impl/ConsentManagementAPIImpl.java:
```diff
@@ -0,0 +1,228 @@
+            }
+
+            ConsentManagementValidationHandler consentManagementValidationHandler = CommonConsentValidationUtil
+                    .getConsentManagementResponseHandler(requestBody.getData().getConsentResourcePath());
```
Is this method name correct? Shouldn't it be `getConsentManagementValidationHandler`?

AthulyaWeerakoon replied:
Yes. Will fix it.

AthulyaWeerakoon replied:
Fixed.

VimukthiRajapaksha commented on components/reference-implementation-openbanking-nextgenpsd2/src/main/java/org/wso2/openbanking/nextgenpsd2/extensions/api/impl/ConsentManagementAPIImpl.java:
```diff
@@ -0,0 +1,228 @@
+            SuccessResponsePreProcessConsentCreation validationResponse;
+
+            if (consentManagementValidationHandler != null) {
+                validationResponse = consentManagementValidationHandler.handleCreation(requestBody);
```
Shall we rename the `validationResponse` variable name to `consentCreationResponse` or something?

AthulyaWeerakoon replied:
But since the purpose of this response is validation, wouldn't that have to be included in the name? maybe `validatedResponse` ?

VimukthiRajapaksha commented on components/reference-implementation-openbanking-nextgenpsd2/src/main/java/org/wso2/openbanking/nextgenpsd2/extensions/api/impl/ConsentManagementAPIImpl.java:
```diff
@@ -0,0 +1,228 @@
+            if (consentManagementValidationHandler != null) {
+                validationResponse = consentManagementValidationHandler.handleCreation(requestBody);
+            } else {
+                JSONObject errorObject = ErrorUtil.getFormattedFailedResponse(404,
```
Consider adding a debug log

AthulyaWeerakoon replied:
Will add.

AthulyaWeerakoon replied:
Added.

VimukthiRajapaksha commented on components/reference-implementation-openbanking-nextgenpsd2/src/main/java/org/wso2/openbanking/nextgenpsd2/extensions/api/impl/ConsentManagementAPIImpl.java:
```diff
@@ -0,0 +1,228 @@
+
+            if (consentManagementValidationHandler != null) {
+                validationResponse = consentManagementValidationHandler.handleRevocation(requestBody);
+            } else {
+                JSONObject errorObject = ErrorUtil.getFormattedFailedResponse(404,
+                        ErrorUtil.constructBerlinError(null, TPPMessage.CategoryEnum.ERROR, null,
+                                ErrorConstants.PATH_INVALID));
+                return Response.ok().entity(errorObject).build();
```
This logic is duplicated in multiple places. If the path is invalid can't we throw an exception from the `getConsentManagementResponseHandler` method?

AthulyaWeerakoon replied:
Will refactor.

AthulyaWeerakoon replied:
Refactored.

AthulyaWeerakoon committed:
Fix minor error handling issue

AthulyaWeerakoon committed:
Minor review fixes

AthulyaWeerakoon commented:
Accidentally closed this PR attempting to rename the branch while the request was still open. The active PR can be found here https://github.com/wso2/reference-implementation-nextgenpsd2/pull/233
